### DISCUSSION TITLE
[Feature] Group by category (and resource)

### DISCUSSION THREAD

**Discussion by patrickunterwegs** [TechbeeAT/jtxBoard#731] [Ideas]
### Discussed in https://github.com/TechbeeAT/jtxBoard/discussions/387

<div type='discussions-op-text'>

<sup>Originally posted by **DickinsonSa** March 20, 2023</sup>
Notes seem to have little way to group when viewing other than status and classification (both marginally useful).

I have dozens of notes with various tags /categories - all of which appear in list view, and the only way to group at the moment is by filtering them individually (time consuming). 

Categories (tags) seem to be a more universally recognised way of grouping (other than notebooks) and yet both these methods are under-utilised for grouping.

Can I put forward for discussion/enhancement please the inclusion of  category in the grouping option for notes?

This would include a 'no category assigned' group also. 

</div>

**Comment by rogercreagh** [comment]
As I said in #387 I would rather see an enhancement to the category-tag filtering to allow options Any | All | None for filtering by multiple tags.
So you could select the tags Red and Apple and get only the red apples (ALL selected tags) where currently you get all red items (including things also tagged as Pears and Cars) and all Apples including the blue and yellow ones. (ANY of the selected tags - which is sometimes useful)

Would this go some way towards meeting your requirement @DickinsonSa? 

I think trying to group by tags when items can have multiple combinations of tags is confusing. Grouping only really works if items can belong to only one group, which is not the case here.

**Comment by DickinsonSa** [comment]
Patrick, it's a different enhancement not connected to what I was thinking of.

But, in light of the confusion that could be caused by grouping multiple tags, I'd just drop my enhancement to suit purity of sticking to the standards.

Forget what I was thinking - which was more akin to notebooks - some better way of displaying hundreds of notes without having to filter.

Enjoy your break.

**Comment by patrickunterwegs** [comment]
Thank you for the discussion, I will not work on that, I'll move the topic to discussion though.

**Comment by storyjesse** [comment]
I've just discovered JTX Board, Thank you for such a fantastic app!

I'm trying to implement a GTD system and JTX is, *almost*, perfect for my needs. The only blockage is that I need the ability to "move" multiple tasks into a "Collection" aka Notebook, the same way I can assign categories/tags to multiple tasks at once. I think this ability might be a good solution to this feature request too since it would then enable Collections/Notebooks to be used the same way Categories/Tags are, with the difference being that each note/journal/task can only be assigned to a single Collection/Notebook. Since notes/journals/tasks can already be grouped by Collections/Notebooks I believe this *interface* addition would potentially solve the original feature request. I also believe that having a *custom* set of "Tags" that have the 'restriction' of one-tag-per-item is actually essential for certain categorization schemes.

Let me explain my use case:

The GettingThingsDone framework starts with **"Capture"** i.e. brainstorm all the tasks/projects/ideas to be done. I might use a Journal for this and create subtasks as I think of them (since a journal has a date which could be useful in the future if I ever want to check back *when* I thought of something.), OR I might just create tasks directly and leave them uncategorised.
Since all tasks/notes/journals need to be in a "Collection/Notebook" I have created a collection/notebook called "GTD" for holding these brainstormed tasks.

The GTD framework's step 2 and 3 are **"Clarify"** and **"Organize"** both of which can often be done together. *Clarifying* involves deciding whether what has been captured is an actionable task, some reference information to be filed, or a complex project that needs further breaking down into actionable tasks. The *Organizing* involves working out WHEN the tasks can/will be done and for 90% of tasks that means assigning them a "Context" i.e. the physical/mental/social/etc circumstances under which the task can be done. [See this Blog post](https://facilethings.com/blog/en/the-organize-stage-of-gtd-explained)

It's important that one *doesn't* categorize tasks at the same time as brainstorming them. I find this is essential because I can get very caught up in categorization and waste lots of time procrastinating about which tasks should be in which category, only to then think of other tasks which require a different categorization scheme.

I initially tried using JTX Board's "Categories" aka Tags for my "Contexts" because I could look through my yet-to-be-organized tasks, select multiple of them at a time and assign a context "Category/Tag" to them. However, I ran into problems when I tried to view my tasks in order to decide what to do next because I couldn't *group* my tasks by "Context" except individually, filtering by one context (category/tag) at a time.

Creating a different Collection/Notebook for each "Context" works well because I can then see a list of my contexts and expand the context I'm currently in, or the one I'm going to be in next and see all the tasks that can be completed in that context. The issue is that I can't easily move tasks between Collection/Notebooks except by doing it individually one tasks at a time which is a time consuming process.

Would it be difficult pragmatically to add a "Collection" section alongside "Categories", "Resources" & "Status" to the "Update entries" menu?

![photo_2023-08-13_13-22-08](https://github.com/TechbeeAT/jtxBoard/assets/2957877/d8c3b4cd-0c07-486a-a1e3-2fefd58b565c)
![photo_2023-08-13_13-22-01](https://github.com/TechbeeAT/jtxBoard/assets/2957877/b84b3ee2-fa66-46df-8482-339eebc25b8e)
![photo_2023-08-13_13-42-16](https://github.com/TechbeeAT/jtxBoard/assets/2957877/478f9089-8f4e-48fe-b54c-ed171edb058a)

**Reply by patrickunterwegs** [reply]
Hey @storyjesse !
Sorry for the late reply, somehow I didn't get a notification on this.

About your request: The option to update the collection through this dialog is already there 😉 I admit it is not clear in your case. You can just scroll the chips horizontally and you will see the other options. So maybe this needs some UI improvement.

I have read through your use case. Basically I still think that categories would be the best choice. You can set up preset categories with colors. Then you could use those preset categories as Kanban-columns. I think this could make a nice overview for your purposes. 

I could have a look how to make a group by category or resource option. The easiest way would be to group by the first category/resource and ignore all others. This would then always work if you only have one category, for multiple categories (if used like tags), it would just take the first...

**Reply by rogercreagh** [reply]
> The easiest way would be to group by the first category/resource and ignore all others. 

NOOOO! See note below. For each item categories appear in the order in which you add them to the item, so the results will be random unless you are super organised and always know which category you will want to group by when creating the item

**Comment by storyjesse** [comment]
Thanks for your response Patrick. I did think of using the kandan view but I have too many categories for this to be practical.
![Screenshot_2023-08-17-08-05-31-41_082f3f9a5723fa65b6a2fa1b83147ead](https://github.com/TechbeeAT/jtxBoard/assets/2957877/deea439a-e926-4759-8dfa-183b21dbfdc0)
![Screenshot_2023-08-17-08-05-49-72_082f3f9a5723fa65b6a2fa1b83147ead](https://github.com/TechbeeAT/jtxBoard/assets/2957877/fe6fda15-285c-44e3-be55-c5896b570972)

**Reply by patrickunterwegs** [reply]
Yeah, I see that this is not very practical. Maybe also this could be improved to be scrollable horizontally and let the elements have a minimum width... I'm just not sure if this would work then properly with the drag and drop functionality...

**Comment by storyjesse** [comment]
You said:
> About your request: The option to update the collection through this dialog is already there 😉 I admit it is not clear in your case. You can just scroll the chips horizontally and you will see the other options. So maybe this needs some UI improvement.

THIS SOLVES MY ISSUE!! 😀

but yes even after you told me about it I couldn't find that functionality because i was swiping sideways in the centre of screen rather than on the headings. I think a UI improvement to scroll the headings when swiping sideways on the body of the screen/window would make this much more intuitive. 😀👍

**Reply by patrickunterwegs** [reply]
Great =)
I already adapted the UI to have a dropdown instead of this horizontally scrollable chips, from 2.5.4 it will be much clearer ;-)

About the "group by category" option: I will do it soon, I started a short survey on this topic on Mastodon, so if you're interested, feel free to participate: https://fosstodon.org/@jtx/110901517568602549

**Reply by rogercreagh** [reply]
I would vote 2 : group by all with entries with multiple categories appearing in multiple groups.

Grouping by only the first doesn't work as the order of categories depends on the order in which they are added per entry - so unless the user remembers to always add the category he will in the future want to group by first the results will be random. 
Suppose you later add a category to an item and want it to be the grouping category for the item - the only way to make it appear as the first category is to remove all existing categories from the item making a (paper?!) note of them then add the new category, then add back all of the others - ugh!

**Reply by patrickunterwegs** [reply]
Yes, the option to group only by the first category would only be a usable option, if the user makes sure that there is only one category. 

Currently the poll favors option 2. But I problem about that came up - having multiple entries with the same unique ID causes a crash in the "LazyList", the lazy list is the most performant way to display scrollable lists with multiple entries... 

As wouldn't say that this is an option, so eventually I might do option 3 because of technical reasons...

**Reply by rogercreagh** [reply]
A thought- does group by Status include custom (x-status) values. Might using x-status for grouping be a better option since each item can have only one status/x-status (ie status is more like what I think of as a category - one per item)

For me a more complete solution remains to enhance the filtering with the ability to combine categories with AND/OR/NOT logic.
1. show items with CatA AND CatB AND CatC
2. Show all items with CatA OR CatB OR CatC (this is the current only option)
3. Show all items NOT assigned to (CatA OR CatB OR CatC) (Invert selection does this but makes the active filter area at the top of the list view very large - another possible improvement if more than half the categories are selected for filtering would be to invert the Active filters display for the category section to just list the ones NOT beng shown?)

Of course there are other logical possibilities CatA AND NOT CatC, Either CatA OR CatC but NOT (CatA AND CatC) etc, but the 3 simple cases cover most requirements, so the only  new thing needed is an AND option added to the 3 dots menu for the category filter, and to display AND NOT OR (or ALL NONE ANY) in the Active filter list. 

I can see that the filter dialogue is already becoming cumbersome - especially if you have a lot of categories - I only have 30, but already I have stopped adding new ones as the categories section is already almost filling the filter screen, and they are shown in a seemingly random (but consistent) order which can make finding them difficult (see issue I'm about to raise on that).

Perhaps for filter sections that could have a lot of entries (primarily Categories and Extended Status) they should expand to a separate page (or accordion section if you have such things in android) which could include the three logic buttons and sort them alphabetically.

The save as preset button might be really useful in such cases. Also would be useful to display the preset filter name in the Active filters area to save space.

**Reply by patrickunterwegs** [reply]
> Perhaps for filter sections that could have a lot of entries (primarily Categories and Extended Status) they should expand to a separate page (or accordion section if you have such things in android) which could include the three logic buttons and sort them alphabetically.

I have changed that a bit, now up to 5 entries are shown and you can click on a "more" button to see all entries


I will check options for the category field to add this ALL, ANY and NONE

**Reply by rogercreagh** [reply]
> I have changed that a bit, now up to 5 entries are shown and you can click on a "more" button to see all entries

Good solution. Thanks. Not in 2.05.02, Hopefully in 2.05.3 not  yet on fDroid here
